src/Functions/FunctionBinaryArithmetic.h

```diff
@@ -1259,4 +1268,167 @@ class FunctionBinaryArithmetic : public IFunction
     bool canBeExecutedOnDefaultArguments() const override { return valid_on_default_arguments; }
 };
 
+
+template <template <typename, typename> class Op, typename Name, bool valid_on_default_arguments = true>
+class FunctionBinaryArithmeticWithConstants : public FunctionBinaryArithmetic<Op, Name, valid_on_default_arguments>
+{
+public:
+    using Base = FunctionBinaryArithmetic<Op, Name, valid_on_default_arguments>;
+    using Monotonicity = typename Base::Monotonicity;
+    static FunctionPtr create(const ColumnWithTypeAndName & left_, const ColumnWithTypeAndName & right_, const Context & context)
+    {
+        return std::make_shared<FunctionBinaryArithmeticWithConstants>(left_, right_, context);
+    }
+    FunctionBinaryArithmeticWithConstants(
+        const ColumnWithTypeAndName & left_, const ColumnWithTypeAndName & right_, const Context & context_)
+        : Base(context_), left(left_), right(right_)
+    {
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t input_rows_count) const override
+    {
+        if (left.column && isColumnConst(*left.column) && arguments.size() == 1)
+        {
+            Block block_with_constant
+                = {{left.column->cloneResized(input_rows_count), left.type, left.name},
+                   block.getByPosition(arguments[0]),
+                   block.getByPosition(result)};
+            Base::executeImpl(block_with_constant, {0, 1}, 2, input_rows_count);
+            block.getByPosition(result) = block_with_constant.getByPosition(2);
+        }
+        else if (right.column && isColumnConst(*right.column) && arguments.size() == 1)
+        {
+            Block block_with_constant
+                = {block.getByPosition(arguments[0]),
+                   {right.column->cloneResized(input_rows_count), right.type, right.name},
+                   block.getByPosition(result)};
+            Base::executeImpl(block_with_constant, {0, 1}, 2, input_rows_count);
+            block.getByPosition(result) = block_with_constant.getByPosition(2);
+        }
+        else
+            Base::executeImpl(block, arguments, result, input_rows_count);
+    }
+
+    bool hasInformationAboutMonotonicity() const override
+    {
+        std::string_view name_ = Name::name;
+        if (name_ == "minus" || name_ == "plus" || name_ == "multiply" || name_ == "divide" || name_ == "intDiv")
+        {
+            return true;
+        }
+        return false;
+    }
+
+    Monotonicity getMonotonicityForRange(const IDataType &, const Field & left_point, const Field & right_point) const override
+    {
+        std::string_view name_ = Name::name;
+        if (name_ == "minus" || name_ == "plus")
+        {
+            return {true, true, true};
```

Comment:
It is not correct because of overflow
```
SELECT 
    18446744073709551614 + number AS arg,
    arg + 1
FROM numbers(2)

┌──────────────────arg─┬─plus(plus(18446744073709551614, number), 1)─┐
│ 18446744073709551614 │                        18446744073709551615 │
│ 18446744073709551615 │                                           0 │
└──────────────────────┴─────────────────────────────────────────────┘

```

Reply:
It should be checked in all other cases too.

Reply:
Ouch. Overflow makes things complicated. I wonder if we should provide saturation arithmetic.